File: 17_Letter_Combinations_of_a_Phone_Number.cpp
#include<iostream>
#include<vector>
#include<string>
using namespace std;

class Solution {
public:
    vector<string> strmap{"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    vector<string> letterCombinations(string digits) {
        if(!digits.length())
            return vector<string>{};
        else{
            vector<string> s{""};
            vector<int> dig;
            for(auto each : digits){
                dig.push_back(each - '0');
            }
            comb(dig, 0, s);
            return s;
        }

    }
    void comb(vector<int> digits, int pos, vector<string> & str){
        if(digits.size() <= pos){
            return;
        }
        auto last = str[str.size()-1];
        str.pop_back();
        for(int i = 0; i < strmap[digits[pos]].length(); i++){
            str.push_back(last + strmap[digits[pos]][i]);
            comb(digits, pos + 1, str);
        }
    }
};

int main(void){
    string digits{"257"};
    auto res = Solution().letterCombinations(digits);
    for(auto each:res){
        cout<<each<<' ';
    }
    return 0;
}

File: 17_Letter_Combinations_of_a_Phone_Number.py
from itertools import product

class Solution:
    def gen(self, strings):
        try:
            for each in strings[0]:
                for every in self.gen(strings[1:]):
                    yield (each,) + every
        except:
            yield ()
    
    def repeat(self, obj, times):
        for _ in range(times):
            yield obj
    
    def letterCombinations(self, digits: str):
        D = ['', 'abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz']
        mapped = list(map(lambda x, y: y[int(x) - 1], digits,[D] * len(digits)))
        return list(map(lambda x: ''.join(x), self.gen(mapped)))

if __name__ == '__main__':
    print(Solution().letterCombinations('23'))